AK: Replace years_to_days_since_epoch by near-instant function

BenWiederhake opened the pull request:
Previously, the function `years_to_days_since_epoch` internally [looped over every single year](https://github.com/SerenityOS/serenity/blob/d26aabff0401690d6dd1332558b40f5cb64e4428/AK/Time.h#L69) between 1970 and the given argument, potentially taking many iterations. I wrote a benchmark that just calls `years_to_days_since_epoch` 100 times. It was very slow:
```
// old:
TestTime(48:48): Completed test 'years_to_days_since_epoch_points' in 59ms
TestTime(48:48): Completed test 'years_to_days_since_epoch_benchmark' in 2324ms
TestTime(48:48): Completed test 'years_to_days_since_epoch_span' in 0ms
```

The entire function (i.e. both loops) can be simply replaced by a single if, about 6 integer divisions, and a dozen additions. It's constant-time. (Not in the cryptographic sense; it just doesn't loop anymore.)
```
// new:
TestTime(49:49): Completed test 'years_to_days_since_epoch_points' in 0ms
TestTime(49:49): Completed test 'years_to_days_since_epoch_benchmark' in 0ms
TestTime(49:49): Completed test 'years_to_days_since_epoch_span' in 0ms
```

The speedup for real-life programs is probably negligible, because *usually* the number of iterations is low and the calls to that function rarely occur in a loop. However, the speedup is very much measurable for specifically-crafted programs like the above benchmark.

nico commented:
Is https://dev.azure.com/SerenityOS/SerenityOS/_build/results?buildId=24833&view=logs&jobId=2351e412-c793-598f-27f8-72c4f2d55ba1&j=2351e412-c793-598f-27f8-72c4f2d55ba1&t=ee375dfa-6b7d-514a-7789-3948d04c0dd3 related or a flake? TestTimeZone is failing and this is changing time code. I think it's probably unrelated, but it's not _obviously_ unrelated.

nico commented on AK/Time.h:
```diff
@@ -66,14 +66,44 @@ constexpr int days_in_year(int year)
+namespace Detail {
+// Consider the integers of the interval [begin, end), and a modulus 'mod'.
+// For each number 'n' in the interval, n % mod may or may not be 0.
+// This function counts how many such 'n' there are in the given interval.
```
"may or may not be 0", "how many such" – not terribly clear from the comment what's going on. Does it count those that may be 0, may not be 0, both?

Either reword ("Counts how many numbers n are in [begin, end) with n % positive_mod == 0"), or remove comment.

BenWiederhake replied:
I love your version, thanks!

nico commented on Tests/AK/TestTime.cpp:
```diff
@@ -274,3 +274,167 @@ TEST_CASE(is_negative)
+
+TEST_CASE(years_to_days_since_epoch_points)
+{
+    auto test_data = Array {
```
nit: I think if you type out the, uh, type, here on the lhs you don't need to repeat it on every line on the rhs I think.

BenWiederhake replied:
I didn't do that while writing the code because then I'll have to hard-code the number "15". But you're right, the committed version should be written this way.

BenWiederhake replied:
This is going to be disgusting, because the compiler wants an initializer_list of exactly one item, namely the array. The way to write that is:
```
    Array<YearAndDays, 15> test_data = { { // Oof
        { 1969, -365 },
        { 1970, 0 },
        { 1971, 365 },
        { 1900, -25567 },
        { 2023, 19358 },
        { 1800, -62091 },
        { 2100, 47482 },
        { 0, -719528 },
        { 4000, 741442 },
        { -10000, -4371953 },
        { 10000, 2932897 },
        { -1000000, -365962028 },
        { 1000000, 364522972 },
        { -5877640, -2147483456 },
        { 5881474, 2147444740 },
    } }; // Oof
```
However, that sounds like it is an issue for another time.

nico reviewed (approved):
Neat!

BenWiederhake commented:
No flakes that I can see, I just did a *very* silly typo, and the tests detected that. Here's what the fix looks like (already squashed in the PR):
```diff
     int days = 365 * (year - 1970);
     int extra_leap_days = 0;
-    days += Detail::mod_zeros_in_range(begin_year, end_year, 4);
-    days -= Detail::mod_zeros_in_range(begin_year, end_year, 100);
-    days += Detail::mod_zeros_in_range(begin_year, end_year, 400);
+    extra_leap_days += Detail::mod_zeros_in_range(begin_year, end_year, 4);
+    extra_leap_days -= Detail::mod_zeros_in_range(begin_year, end_year, 100);
+    extra_leap_days += Detail::mod_zeros_in_range(begin_year, end_year, 400);
     return days + extra_leap_days * leap_sign;
```
D'oh!

BenWiederhake commented:
Changes since first push:
- Completely avoided signed-integer overflow in all cases, in contrast to the old implementation (new commit)
- Fixed a silly typo introduced by that
- Improved the comment/explanation on the crucial `mod_zeros_in_range` function (thanks @nico!)
- Use a nicer type in the test (thanks @nico!)
- Remove a FIXME from LibJS, because it looks incorrect to me, and the old reasoning does not apply anymore.
- Use a more appropriate function in the Kernel

BenWiederhake commented:
Changes since last push:
- Fixed really silly typo (sorry)

BenWiederhake commented:
Changes since last push:
- Several attempts later, I think I finally got *actually* rid of integer overflow.

In other news, rebuild times are surprisingly high, and way too many things include AK/Time.h.

trflynn89 commented on Tests/AK/TestTime.cpp:
```diff
@@ -274,3 +274,130 @@ TEST_CASE(is_negative)
+        190657, 191023, 191388, 191753, 192118, 192484, 192849, 193214,
+        193579, 193944, 194309, 194674, 195039, 195405,
+    };
+    // clang-format off
```
`clang-format on`

BenWiederhake replied:
Derp! You're right.

BenWiederhake commented:
Changes since last push:
- Fixed wrong clang-format off/on tag (thanks @trflynn89!)
- Rebased for convenience

trflynn89 commented on Userland/Libraries/LibJS/Runtime/Date.cpp:
```diff
@@ -483,8 +483,7 @@ double make_day(double year, double month, double date)
 
-    // FIXME: We are avoiding AK::years_to_days_since_epoch here because it is implemented by looping over
-    //        the range [1970, ym), which will spin for any time value with an extremely large year.
+    // We cannot use AK::{years_to_,}days_since_epoch, because it might change the values slightly.
```
Let's leave the FIXME in this comment. I remember when I wrote this that it felt like a shame that we couldn't use `years_to_days_since_epoch`. If the values are different, there's probably something worth looking into in LibJS.

BenWiederhake replied:
I didn't try replacing it, so I'm not entirely certain that the values would differ.

However, it seems that all of the rounding and overflow stuff is hardcoded in the spec, so replacing it by a call to `{years_to_,}days_since_epoch` means that it's harder to verify that we satisfy the spec.

If you want to keep the FIXME in, how do you suggest phrasing it? In particular, I don't think there is anything left to fix here: We adhere to the spec. The comment explains why we can't do it the "better" way.

trflynn89 replied:
Essentially step 8 would here would become:
```js
    // 8. Find a finite time value t such that YearFromTime(t) is ym and MonthFromTime(t) is mn and DateFromTime(t) is 1𝔽; but if this is not possible (because some argument is out of range), return NaN.
    if (!AK::is_within_range<int>(ym) || !AK::is_within_range<int>(mn + 1))
        return NAN;

    auto t = (years_to_days_since_epoch(static_cast<int>(ym)) + day_of_year(static_cast<int>(ym), static_cast<int>(mn) + 1, 1)) * ms_per_day;
    VERIFY(year_from_time(t) == ym);
    VERIFY(month_from_time(t) == mn);
    VERIFY(date_from_time(t) == 1);
```

I think this would be better, because what I originally wrote is mixing `Date.prototype` math with `AK::Time` math, whereas this approach sticks with `AK::Time` math. It just wasn't feasible with the slow `for`-loops before.

Sidenote, when the spec says "Find a finite value" or something similar without actually telling you *how* to do it, we're in implementation-defined territory. We can/should verify that any invariant holds true (such as the `VERIFY`s written above), but beyond that, we're simultaneously on our own and can do we want (within reason).

trflynn89 replied:
(To be clear, I'm not asking you to be responsible for that change here, just to leave the FIXME so we can do it another time)

trflynn89 replied:
Took a hot minute to remember this, but basically the FIXME is reverting 032664332be6e1d42567165e9ffb623faf2b9a10.

BenWiederhake replied:
That looks good! I'll use that, thank you!

trflynn89 commented on Tests/AK/TestTime.cpp:
```diff
@@ -274,3 +274,130 @@ TEST_CASE(is_negative)
+    }
+}
+
+TEST_CASE(years_to_days_since_epoch_benchmark)
```
If this is meant to only be a benchmark, we have `BENCHMARK_CASE` for this.

BenWiederhake replied:
Good idea! I forgot that it exists, thanks!

BenWiederhake commented:
Changes since last push:
- Use `BENCHMARK_CASE` instead of `TEST_CASE` for benchmark (thanks @trflynn89!)
- Rebased for convenience

Output after the first commit (which introduces only the benchmark, not the speedup):
```
Running test 'years_to_days_since_epoch_points'.
Completed test 'years_to_days_since_epoch_points' in 6ms
Running benchmark 'years_to_days_since_epoch_benchmark'.
Completed benchmark 'years_to_days_since_epoch_benchmark' in 296ms
```

Output after the speedup:
```
Running test 'years_to_days_since_epoch_points'.
Completed test 'years_to_days_since_epoch_points' in 0ms
Running benchmark 'years_to_days_since_epoch_benchmark'.
Completed benchmark 'years_to_days_since_epoch_benchmark' in 0ms
```

trflynn89 commented:
@BenWiederhake While testing this out, I came across a year that has different results now than it did before (not due to overflow): `-999999` (test262 loves this year, the wine must've been good or something).

Before:
```
years_to_days_since_epoch(-999999) == -365961662
```

After:
```
years_to_days_since_epoch(-999999) == -365961663
```

It seems the former is correct based on some test262 expectation?

BertalanD commented:
FWIW, I found another interesting implementation of this algorithm, maybe you could verify this one against that: https://howardhinnant.github.io/date_algorithms.html#days_from_civil

trflynn89 commented:
> Oh, and this old issue of mine might be somewhat relevant: #12729

I thought this all looked very familiar, for more discussion/test cases, see #13138

BenWiederhake commented:
Turns out the logic is right, it's just that I tripped up on C++ division logic. I still can't wrap my head around the question why oh WHY C has integer division rounding toward zero. This makes everything so more complicated!

Specifically, `AK::Detail::mod_zeros_in_range(-3, 3, 4)` should be 1, but right now computes 2, although there aren't two zeros. I'm working on it, just wanted to give an update.

BenWiederhake commented:
Changes since last push:
- Marked #12729 for auto-closure as suggested by @BertalanD.
- Fixed division logic and added test. Division-rounding-to-negative-inf is surprisingly non-trivial in C++. This resolves the issue of years_to_days_since_epoch(-999999). The PR now also contains a test for specifically that value.
- Converted the LibJS changes to a revert commit, as suggested by @trflynn89. Thanks!

trflynn89 commented:
> Marked LibC: localtime is very slow for large time_t values #12729 for auto-closure as suggested by @BertalanD.

That's not fixed here, is it? `time_to_tm` is still using the slow `for`-loop.

BenWiederhake commented:
> `time_to_tm`

We seriously have two implementations of that? Oh god no.

Can I just replace the code duplication by a call to AK::Time, what do you think?

trflynn89 commented:
> Can I just replace the code duplication by a call to AK::Time, what do you think?

For sure, we are free to call into `AK::Time` from LibC.

BenWiederhake commented:
> For sure, we are free to call into `AK::Time` from LibC.

Reading a bit more closely, time_to_tm is the inverse function, so it does "the opposite" of years_to_days_since_epoch. In fact, tm_to_time automatically got the speedup, because it just forwards to years_to_days_since_epoch.

*Theoretically*, one might multiply/divide by 365.2425, which always gives a result that is off by at most one. That would give an excellent initial guess, which can then be refined through years_to_days_since_epoch. However, I'm not going to implement that, especially not in this PR.
